pkg/plugins/runtime/k8s/controllers/gateway_instance_controller.go

```diff
@@ -0,0 +1,229 @@
+	gatewayInstance := &mesh_k8s.GatewayInstance{}
+	if err := r.Get(ctx, req.NamespacedName, gatewayInstance); err != nil {
+		if kube_apierrs.IsNotFound(err) {
+			return kube_ctrl.Result{}, nil
+		}
+
+		return kube_ctrl.Result{}, err
+	}
+
+	deployment, err := r.createOrUpdateDeployment(ctx, gatewayInstance)
+	if err != nil {
+		return kube_ctrl.Result{}, errors.Wrap(err, "unable to create Deployment for Gateway")
+	}
+
+	svc, err := r.createOrUpdateService(ctx, gatewayInstance)
+	if err != nil {
+		return kube_ctrl.Result{}, errors.Wrap(err, "unable to create Service for Gateway")
+	}
+
+	updateStatus(gatewayInstance, svc, deployment)
+
+	if err := r.Client.Status().Update(ctx, gatewayInstance); err != nil {
+		return kube_ctrl.Result{}, errors.Wrap(err, "unable to update GatewayInstance status")
+	}
+
+	return kube_ctrl.Result{}, nil
+}
+
+func k8sSelector(name string) map[string]string {
+	return map[string]string{"app": name}
+}
+
+func (r *GatewayInstanceReconciler) createOrUpdateService(
+	ctx context.Context,
+	gatewayInstance *mesh_k8s.GatewayInstance,
+) (*kube_core.Service, error) {
+	service := &kube_core.Service{
+		ObjectMeta: kube_meta.ObjectMeta{
+			Namespace: gatewayInstance.Namespace,
+			Name:      gatewayInstance.Name,
+		},
+	}
+
+	gateway := match.Gateway(r.ResourceManager, func(selector mesh_proto.TagSelector) bool {
+		return selector.Matches(gatewayInstance.Spec.Tags)
+	})
+
+	if gateway == nil {
+		return nil, fmt.Errorf("no valid Gateway found")
+	}
+
+	if _, err := kube_controllerutil.CreateOrUpdate(ctx, r.Client, service, func() error {
+		var ports []kube_core.ServicePort
+
+		for _, listener := range gateway.Spec.GetConf().GetListeners() {
+			ports = append(ports, kube_core.ServicePort{
+				Name:     strconv.Itoa(int(listener.Port)),
+				Protocol: kube_core.ProtocolTCP,
+				Port:     int32(listener.Port),
+			})
+		}
+
+		service.Spec = kube_core.ServiceSpec{
+			Selector: k8sSelector(gatewayInstance.Name),
+			Ports:    ports,
+			Type:     gatewayInstance.Spec.ServiceType,
+		}
+
+		err := kube_controllerutil.SetControllerReference(gatewayInstance, service, r.Scheme)
+		return errors.Wrap(err, "unable to set Service's controller reference to GatewayInstance")
+	}); err != nil {
+		return nil, errors.Wrap(err, "unable to create or update Service for GatewayInstance")
+	}
+
+	return service, nil
+}
+
+func (r *GatewayInstanceReconciler) createOrUpdateDeployment(
+	ctx context.Context,
+	gatewayInstance *mesh_k8s.GatewayInstance,
+) (*kube_apps.Deployment, error) {
+	ns := kube_core.Namespace{}
+	if err := r.Client.Get(ctx, kube_types.NamespacedName{Name: gatewayInstance.Namespace}, &ns); err != nil {
+		return nil, errors.Wrap(err, "unable to get Namespace for gateway")
+	}
+
+	deployment := &kube_apps.Deployment{
+		ObjectMeta: kube_meta.ObjectMeta{
+			Namespace: gatewayInstance.GetNamespace(),
+			Name:      gatewayInstance.GetName(),
```

Comment:
When we create the child objects we ought to use `generateName` so that we can avoid name collisions with objects that users create.

This looks like the recommended way to find the objects that an object owns:

https://github.com/kubernetes-sigs/kubebuilder/blob/master/docs/book/src/cronjob-tutorial/testdata/project/controllers/cronjob_controller.go#L120-L124

Reply:
Cool, I didn't know `GenerateName` existed, I'm setting `GenerateName` now instead of `Name`

Reply:
I think this requires adjustment to the `CreateOrUpdate` calls.

Reply:
@jpeach Please take a look at the newest commit